Prevent auth from starting when an incompatible major upgrade is detected

vapopov opened the pull request:
Added KV for storing teleport version in backend. If the current major version of teleport bigger than previous one more than one we must refuse to start.

Closes #41782

changelog: Teleport auth prevents upgrade major version more than one

vapopov committed:
Teleport auth should refuse to start when jumping major versions (#41782)

vapopov committed:
Merge remote-tracking branch 'origin/master' into vapopov/major-version-check

vapopov committed:
linter fixes

rosstimothy reviewed (commented):
Couple of high level thoughts:
1) I don't think we need the version quasi resource in api/types. This should only be internal to Auth and not something exposed to users.
2) We should use the semver package instead of regular expressions to parse versions.
3) ValidateAndUpdateTeleportVersion is currently very racy. Multiple Auth processes that start at the same time will both see no initial version in the backend and overwrite the others version. We should make use of ConditionalUpdate or AtomicWrite to guarantee that we are always acting on the most recent data stored in the backend and cannot accidentally overwrite the version of another process.
4) The error returned when the differences in versions should prevent starting Auth should be more informative and actionable. Let's make sure the users know that what they are trying to do is not supported, and inform them of the proper major version to upgrade to in an effort to reduce support load.
5) We should also consider in the future that an absence of a version in the backend might also mean the upgrade should not proceed. If this is introduced in v17, any updates from < v17 to >= v18 will not have a persisted version but would be an attempt to skip major versions.

zmb3 commented on lib/services/local/version.go:
```diff
@@ -0,0 +1,81 @@
+	"context"
+
+	"github.com/gravitational/trace"
+	"github.com/sirupsen/logrus"
```
Can we use slog for new code so we don't have to convert it in the future?

zmb3 commented on lib/services/local/version.go:
```diff
@@ -0,0 +1,81 @@
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+package local
```
Sure, but we're not trying to make this bulletproof. You could make the same argument for the backend - "couldn't the user just remove the version entry from the backend to violate the upgrade guidance?"

r0mant replied:
I think keeping it in local process storage would also not work in cases when auth runs in Kubernetes and pods recycle (which is every upgrade basically).

r0mant replied:
Actually, ignore my previous comment, we use secrets for local storage in Kubernetes so it would work. In fact, it looks like @fspmarshall added something called [InitialLocalVersion](https://github.com/gravitational/teleport/blob/c44530d985cf07ad4fa849b30281d09fc6ca2813/lib/auth/state/state.go#L90) to process storage a few weeks ago in https://github.com/gravitational/teleport/pull/42189. Is this something we can piggy-back on, or add the version we need to record alongside it? @fspmarshall Is "InitialLocalVersion" supposed to be updated when teleport upgrades?

vapopov replied:
@r0mant I've launched latest v17, so we keep initial version `"initial_local_version":"15.4.5"`
what about to extend this state object to add version history field as array to keep track all version launch (as usually migration logic does)

r0mant replied:
I don't know if we need to keep the version history necessarily but adding another process state field for tracking the latest and current version (unless we can reuse initial_local_version) seems reasonable.

vapopov commented on lib/services/local/version.go:
```diff
@@ -0,0 +1,81 @@
+}
+
+const (
+	versionPrefix = "version"
```
What about moving it to auth scope like `backend.Key(authPrefix, versionPrefix)`

vapopov committed:
code review changes:
- replaced to use process storage and state itme
- regexp replaced with semver
- additional check for v18 for empty version record in storage

r0mant commented on lib/service/version.go:
```diff
@@ -0,0 +1,95 @@
+
+var (
+	// errMajorVersionUpgrade default error for restricting upgrade major version.
+	errMajorVersionUpgrade = errors.New(`Teleport supports upgrade only one major version behind. ` +
```
As mentioned below, I don't think you need this variable. I'd just `return trace.BadParameter(<message>)` inline.

r0mant commented on lib/service/version.go:
```diff
@@ -0,0 +1,95 @@
+	}
+
+	if st.Spec.LocalVersion == "" {
+		if currentMajor >= majorVersionConstraint {
```
Do we ensure that new clusters have the local version set during the first start?

r0mant commented on lib/service/connect.go:
```diff
@@ -552,6 +552,7 @@ func (process *TeleportProcess) firstTimeConnect(role types.SystemRole) (*Connec
 		Spec: state.StateSpecV2{
 			Rotation:            ca.GetRotation(),
 			InitialLocalVersion: teleport.Version,
+			LocalVersion:        teleport.Version,
```
Hmm I'm not sure this is the correct place to ensure auth server has this version set in its local state upon first start. I think this code runs on agents. You might wanna look at Init in https://github.com/gravitational/teleport/blob/master/lib/auth/init.go.

r0mant commented on lib/service/version.go:
```diff
@@ -0,0 +1,95 @@
+				currentVersion, currentMajor-1)
+		}
+
+		st.Spec.LocalVersion = currentVersion
+		if err := process.storage.WriteState(types.RoleInstance, *st); err != nil {
+			return trace.Wrap(err)
+		}
+		return nil
```
I think the logic would be slightly cleaner if you moved this out of the loop (or rather just removed it since it's already done below) and put the version check under "else". So it would look like this:

```go
if st.Spec.LocalVersion == "" {
  // check version constraint
} else {
  // check existing local version
}

st.Spec.LocalVersion = currentVersion
if err := process.storage.WriteState(types.RoleInstance, *st); err != nil {
	return trace.Wrap(err)
}
return nil
```

r0mant commented on lib/service/version.go:
```diff
@@ -0,0 +1,95 @@
+		return trace.Wrap(err)
+	}
+	if currentMajor-localMajor > 1 {
+		return trace.Wrap(errMajorVersionUpgrade, "Teleport version %s is going to upgrade to %s, but first you need to upgrade to version %d.x.x at least",
```
Couple of comments:

- Here you don't need to wrap an error, I would just return a single `trace.BadParameter(<error message>)` so you don't have to define a separate `errMajorVersionUpgrade`.
- This error message is better than the previous one but I would tweak it a bit more. For example:

```suggestion
			return trace.BadParameter("Unsupported upgrade path detected: from %v to %v. Teleport supports direct upgrades to the next major version only. Please upgrade your cluster to version %d.x.x first. See compatibility guarantees for details: https://goteleport.com/docs/upgrading/overview/#component-compatibility.",
```

r0mant commented on lib/service/version.go:
```diff
@@ -0,0 +1,95 @@
+		return trace.Wrap(err)
+	}
+
+	st, err := process.storage.GetState(process.GracefulExitContext(), types.RoleInstance)
```
Nit: `st` -> `state`. Prefer full variable names, shorthands make code harder to read. It's a bit against [style guide](https://go.dev/wiki/CodeReviewComments#variable-names) but I think it more applies to loop variables, for vars that are used throughout the function we prefer reasonably descriptive names.

r0mant commented on lib/service/version.go:
```diff
@@ -0,0 +1,95 @@
+}
+
+// getMajorVersion parses string to fetch major version number.
+func getMajorVersion(version string) (int64, error) {
```
Consider moving this to https://github.com/gravitational/teleport/blob/master/lib/utils/ver.go. There's already [MajorSemver](https://github.com/gravitational/teleport/blob/8d102a37881519a5da8d44f0d2e7d521c2f47973/lib/utils/ver.go#L77) function but it returns a string. Maybe move it there and call `MajorSemverNum` or something?

r0mant commented on lib/service/version.go:
```diff
@@ -0,0 +1,95 @@
+	}
+
+	st, err := process.storage.GetState(process.GracefulExitContext(), types.RoleInstance)
+	if trace.IsNotFound(err) {
+		// If the Instance state has not been found, it means this is the first launch,
+		// and we don't need to check the version since this is the initial launch.
+		return nil
```
I think we need to test a scenario when local process state gets removed, for whatever reason (shouldn't normally happen but people can get their clusters in all sorts of weird states). What would happen in this case? This will not be considered "first time start" because that is determined based on the presence of CA IIRC, will we re-init the process state with correct version? I'm mostly worried about scenario where it would get into some state which we'll have to recover manually which we need to avoid.

vapopov replied:
For instance if we remove process database completely or specifically this `/states/instance/state` item, first we should go in this version check so `/states/instance/state` not created yet, this item managed by `process.connect` and going to be created after validation. Currently looking to in integration tests, to cover this case

Worst case scenario might be, if the user manually change `local_version` of `/states/instance/state` item in database

vapopov replied:
Was testing couple of scenarios with integration tests, found that in configuration when we enable only auth service
```golang
	cfg.Auth.Enabled = true
	cfg.Proxy.Enabled = false
	cfg.SSH.Enabled = false
```
We never register the instance state in the process storage
https://github.com/gravitational/teleport/blob/dc1199636558f85ed275512175a12207a8034d44/lib/service/service.go#L2564-L2574
this leads to skipping version check in auth service

r0mant replied:
@fspmarshall Any advice here?

r0mant reviewed (commented):
In addition to the comments, can you add test coverage to the PR?

zmb3 commented:
@vapopov I've changed the title a bit so that it describes what the change does, not what the change _should_ do.

I also removed the corresponding issue from the PR title. It's better to link the issue in the PR body, not the title.

vapopov committed:
code review changes:
- add integration test
- store latest version in separate key

zmb3 commented on integration/version_test.go:
```diff
@@ -0,0 +1,258 @@
+	processStorage, err := storage.NewProcessStorage(ctx, filepath.Join(authDir, teleport.ComponentProcess))
+	require.NoError(t, err)
+
+	authAddr, err := getFreeListenAddr()
```
I would avoid `getFreeListenAddr()` if you can. It's a flawed approach.

zmb3 commented on integration/version_test.go:
```diff
@@ -0,0 +1,258 @@
+}
+
+// TestVersionCheckUpgradeFailed tests version upgrade restriction with incompatible major upgrade.
+func TestVersionCheckUpgradeFailed(t *testing.T) {
```
Looks like these 3 tests share a lot of common code. Can we turn them into table tests?

zmb3 commented on lib/auth/storage/storage.go:
```diff
@@ -45,6 +45,10 @@ const (
 	idsPrefix = "ids"
+	// teleportPrefix is a key prefix to store internal data
+	teleportPrefix = "teleport"
+	// latestVersion is a key for storing latest version of teleport
```
"latest" is a big ambiguous here. Is it the latest released version, the latest version running in the cluster, etc..

Should we call this `lastKnownVersion` or something similar?

zmb3 commented on lib/service/version.go:
```diff
@@ -33,63 +31,47 @@ const (
 	majorVersionConstraint = 18
```
```suggestion
	// TODO(vapopov): DELETE IN 19
	majorVersionConstraint = 18
```

zmb3 commented on lib/service/version.go:
```diff
@@ -33,63 +31,47 @@ const (
+			return trace.BadParameter("Unsupported upgrade path detected: to %v. "+
+				"Teleport supports direct upgrades to the next major version only.\n Please upgrade "+
+				"your cluster to version %d.x.x first. See compatibility guarantees for details: "+
+				"https://goteleport.com/docs/upgrading/overview/#component-compatibility.",
```
I think this error message is incorrect.

If you were on Teleport 15 and you tried to upgrade straight to Teleport 18, wouldn't it say "Please upgrade your cluster to version 17.x.x first" ?? 

If so, that's not right, we should be telling them to upgrade to v16 first.

vapopov replied:
in this specific case we couldn't know exact version of previous setup, I would probably change the message with example from docs instead of this version suggestion

zmb3 commented on lib/utils/ver.go:
```diff
@@ -82,6 +82,15 @@ func MajorSemver(version string) (string, error) {
+func MajorVersion(version string) (int64, error) {
+	v, err := semver.NewVersion(version)
+	if err != nil {
+		return 0, trace.Wrap(err, "unsupported version format: %q", version)
```
```suggestion
		return 0, trace.BadParameter(err, "unsupported version format: %q", version)
```

vapopov committed:
code review changes:
- table tests
- error message

zmb3 commented on lib/auth/storage/storage.go:
```diff
@@ -203,6 +207,28 @@ func (p *ProcessStorage) WriteIdentity(name string, id state.Identity) error {
 	return trace.Wrap(err)
 }
 
+// GetTeleportVersion reads the latest teleport version.
```
```suggestion
// GetTeleportVersion reads the last known Teleport version from storage.
```

zmb3 commented on lib/auth/storage/storage.go:
```diff
@@ -203,6 +207,28 @@ func (p *ProcessStorage) WriteIdentity(name string, id state.Identity) error {
+		Key:   backend.Key(teleportPrefix, lastKnownVersion),
+		Value: []byte(version),
+	}
+	_, err := p.stateStorage.Put(context.TODO(), item)
```
Do you plan to address this TODO? If not, use `context.Background()`.

zmb3 commented on lib/auth/storage/storage.go:
```diff
@@ -203,6 +207,28 @@ func (p *ProcessStorage) WriteIdentity(name string, id state.Identity) error {
+	return string(item.Value), nil
+}
+
+// WriteTeleportVersion writes the latest teleport version to the backend.
```
As mentioned, I think "latest" is confusing here.

zmb3 commented on lib/service/version.go:
```diff
@@ -0,0 +1,79 @@
+		if err != nil {
+			return trace.Wrap(err)
+		}
+		if currentMajor-localMajor > 1 {
```
Hmm, any way we can simplify this function? We're like 5 levels deep in if statements at this point.

vapopov committed:
code review changes:
- minor changes

vapopov commented on lib/auth/storage/storage.go:
```diff
@@ -203,6 +207,28 @@ func (p *ProcessStorage) WriteIdentity(name string, id state.Identity) error {
+}
+
+// WriteTeleportVersion writes the last known Teleport version to the storage.
+func (p *ProcessStorage) WriteTeleportVersion(version string) error {
+	item := backend.Item{
+		Key:   backend.Key(teleportPrefix, lastKnownVersion),
+		Value: []byte(version),
+	}
+	_, err := p.stateStorage.Put(context.Background(), item)
```
yes, I made similar as state written

vapopov committed:
code review changes:
- reuse context for write operation

r0mant commented on lib/service/version.go:
```diff
@@ -0,0 +1,82 @@
+	if err != nil {
+		return trace.Wrap(err)
+	}
+	if currentMajor-localMajor > 1 {
```
It just occurred to me: we should probably prevent accidental downgrades by more than 1 major apart too.

r0mant commented on integration/version_test.go:
```diff
@@ -0,0 +1,146 @@
+			signal := make(chan struct{})
+			go launchFunc(ctx, *authCfg, authRunErrCh, signal)
+
+			timeout := time.After(time.Second * 30)
+			select {
+			case err := <-authRunErrCh:
+				if test.expectError {
+					assert.Error(t, err)
+				} else {
+					require.NoError(t, err)
+				}
+			case <-timeout:
+				t.Fatal("timed out waiting for fetching identity")
+			case <-signal:
+			}
```
We have a pretty major problem with flaky tests in Teleport, due to tests starting and waiting for various services which we largely have under control now but still, is there a way to make this test not depend on time?

For example, why do we need to start auth? Can we just instantiate auth of version X, call "initAuthService" on it, then start another auth of version X+1, init, make sure it works, then another of version X+3, init, make sure it fails, something like that? There would be no potential flakiness and the test would probably be simpler as well.

vapopov replied:
I need to additionally ensure that state for `RoleAdmin` created and this logic might be changed in time in initialization flow, thats why initially I choose to use integration test for such case. I check this state to identify if we have new install with empty database rather than existing one.

For instance if I add tests which trigger `initAuthService` only, we will never cover the case with `majorVersionConstraint` (this case covers situation: let say we have version 15, we know that storing information about version added in v17, and we are trying to update to v18, so if we check that we have any version in storage means that we have older version that <v17)

vapopov replied:
I rewrite the test to use `initAuthService` instead

r0mant reviewed (commented):
I think we're getting there but left a few more comments.

vapopov commented on lib/service/version.go:
```diff
@@ -0,0 +1,82 @@
+	if trace.IsNotFound(err) {
+		// We have to ensure that the process database is newly created before
+		//applying the majorVersionConstraint check.
+		_, err := process.storage.GetState(process.GracefulExitContext(), types.RoleAdmin)
```
By default we create state `RoleAdmin` only (in setup where we configure Teleport with auth service and disable everything else)

vapopov commented on lib/auth/storage/storage.go:
```diff
@@ -203,6 +207,28 @@ func (p *ProcessStorage) WriteIdentity(name string, id state.Identity) error {
 
+// GetTeleportVersion reads the last known Teleport version from storage.
+func (p *ProcessStorage) GetTeleportVersion(ctx context.Context) (string, error) {
+	item, err := p.stateStorage.Get(ctx, backend.Key(teleportPrefix, lastKnownVersion))
```
@fspmarshall proposed to make it global in context of the roles, should I change it?

r0mant replied:
Probably fine as is.

vapopov commented on lib/service/version.go:
```diff
@@ -0,0 +1,82 @@
+		if !trace.IsNotFound(err) && err != nil {
+			return trace.Wrap(err)
+		}
+		if currentMajor >= majorVersionConstraint && !trace.IsNotFound(err) {
```
yes, this logic is related to identifying the first launch, we have something similar for `firstTimeConnect` but can't reuse it. Another option might be to add to the process backend interface a method to check if we have any items stored in the backend

vapopov committed:
code review changes:
- simplify integration test
- add downgrade check
- minor changes

vapopov commented on lib/service/version.go:
```diff
@@ -0,0 +1,81 @@
+
+// validateAndUpdateTeleportVersion validates that the major version persistent in the backend
+// meets our upgrade compatibility guide.
+func (process *TeleportProcess) validateAndUpdateTeleportVersion(currentVersion *semver.Version, firstTimeStart bool) error {
```
I wanted to have this check before backend storage init

https://github.com/gravitational/teleport/blob/fbc4636da8d72b7d21ac572ea0282289d0dfdbf9/lib/service/service.go#L1781-L1790

vapopov replied:
@rosstimothy I've moved logic to `auth.Init`

I've checked migration functionality in `pgbk` and `lite` backend drivers happens in `process.initAuthStorage()`, as I understood we don't plan to extend schemas or do any upgrades of existing tables so we should be good with check in `auth.Init`

vapopov committed:
code review changes:
- move logic to the auth init

r0mant commented on lib/auth/version.go:
```diff
@@ -0,0 +1,89 @@
+
+// validateAndUpdateTeleportVersion validates that the major version persistent in the backend
+// meets our upgrade compatibility guide.
+func validateAndUpdateTeleportVersion(
```
Should we have some kind of a hidden escape hatch? Like an (undocumented) environment variable that skips this check. Would be useful for development purposes.

vapopov replied:
I might add something similar as we already have in `InitConfig`:
```
	// SkipPeriodicOperations turns off periodic operations
	// used in tests that don't need periodic operations.
	SkipPeriodicOperations bool
```
so this value going to be accessible only from code

r0mant replied:
@vapopov Having it in the code only won't help much for development IMO. The idea is that when working on multiple different versions of Teleport you might need to switch back-and-forth between them on the same cluster, e.g. from v14 to v16 and vice versa to test something. That's been my experience before at least - and even though for production scenarios it is not supported upgrade/downgrade path, for development it works fine most of the time. So the idea is that there should be something devs can set in their environment to bypass this check.

r0mant commented on lib/auth/init.go:
```diff
@@ -76,6 +77,9 @@ type InitConfig struct {
 	Backend backend.Backend
 
+	// ProcessStorage is a backend for local process
+	ProcessStorage *storage.ProcessStorage
```
Need to update lib/auth/auth.go:NewServer now to make sure this field is not nil to avoid potential panics.

vapopov replied:
I've added PR with tests fix: https://github.com/gravitational/teleport.e/pull/4594

r0mant reviewed (commented):
mostly lgtm with a couple remaining comments

vapopov committed:
code review changes:
- add development env to skip version check
- init the process storage if it is not set

vapopov commented on lib/auth/storage/storage.go:
```diff
@@ -203,6 +207,28 @@ func (p *ProcessStorage) WriteIdentity(name string, id state.Identity) error {
 }
 
+// GetTeleportVersion reads the last known Teleport version from storage.
+func (p *ProcessStorage) GetTeleportVersion(ctx context.Context) (string, error) {
```
I intentionally left the string not to have dependency in storage from another package

vapopov commented on lib/auth/version.go:
```diff
@@ -0,0 +1,98 @@
+	firstTimeStart bool,
+) error {
+	if skip := os.Getenv(skipVersionUpgradeCheckEnv); skip != "" {
+		return nil
```
If we want to test something with major version upgrade and disable this check, by constraint we anyway shouldn't  persist the version, in other cases we don't need to disable it

vapopov commented on lib/auth/init.go:
```diff
@@ -76,6 +77,9 @@ type InitConfig struct {
 	Backend backend.Backend
 
+	// ProcessStorage is a backend for local process
+	ProcessStorage *storage.ProcessStorage
```
I would add other methods to the interface as well. if the process storage constructor doesn't initialize the backend internally but rather accepts it as an argument. In such a case, we can mock the backend storage instead

vapopov committed:
code review changes:
- add interface for local process storage
- use semver in storage instead

vapopov committed:
change interface to version storage

vapopov commented on lib/auth/auth_test.go:
```diff
@@ -84,10 +86,11 @@ import (
-	a           *Server
-	mockEmitter *eventstest.MockRecorderEmitter
+	bk             backend.Backend
+	processStorage *storage.ProcessStorage
```
@rosstimothy in-memory implementation shouldn't work?

vapopov committed:
add fake version storage for other tests

vapopov commented on lib/auth/helpers.go:
```diff
@@ -1128,6 +1131,26 @@ func (t *TestTLSServer) Stop() error {
+}
+
+// FakeTeleportVersion creates fake version storage.
+func FakeTeleportVersion() VersionStorage {
```
misprint, I was changing from in-memory implementation

ravicious commented on lib/auth/version.go:
```diff
@@ -0,0 +1,91 @@
+		return trace.Wrap(err)
+	}
+
+	if currentVersion.Major-lastKnownVersion.Major > 1 {
+		return trace.BadParameter("Unsupported upgrade path detected: from %v to %v. "+
+			"Teleport supports direct upgrades to the next major version only.\n Please upgrade "+
+			"your cluster to version %d.x.x first. See compatibility guarantees for details: "+
+			"https://goteleport.com/docs/upgrading/overview/#component-compatibility.",
+			lastKnownVersion, currentVersion.String(), lastKnownVersion.Major+1)
+	}
```
I just did what Roman described in another thread (https://github.com/gravitational/teleport/pull/43520/files#r1667060081): I switched to branch/v14, ran `make build/teleport`, started teleport and then switched back to master. Now I cannot start my cluster from master on 4aa48dbcbf0, because I keep getting this error:

```
initialization failed
	Unsupported upgrade path detected: from 14.3.21 to 17.0.0-dev. Teleport supports direct upgrades to the next major version only.
 Please upgrade your cluster to version 15.x.x first. See compatibility guarantees for details: https://goteleport.com/docs/upgrading/overview/#component-compatibility.
```

So I downloaded 15.4.9 and 16.1.0, started them one after the other, but it didn't fix the problem. How does it know that my last version was 14.3.21 but cannot figure out that I ran 15.4.9 and 16.1.0 in the meantime? 🤔 Do I need to trigger a particular code path while 15.x and 16.x are running?

vapopov replied:
@ravicious you need to switch to branch/v15, because seems like 15.4.9 
 https://github.com/gravitational/teleport/pull/44051 has no these changes in this release, also you can use `TELEPORT_UNSTABLE_SKIP_VERSION_UPGRADE_CHECK=yes` to skip this check
